### ISSUE TITLE
Hangs in get-file and delete-all on s3

### ISSUE THREAD

**Issue by jdoliner** [pachyderm/pachyderm#717]
Users have reported experiencing hangs with repeated calls to `delete-all` and `get-file` when running on s3. My first guess is that it has something to do with rate limiting. We need to develop more test cover on this, see #648. Looking at the code for our block server here https://github.com/pachyderm/pachyderm/blob/master/src/server/pfs/server/obj_block_api_server.go#L160 we can see that we're only doing retries on `GetBlock` probably the first thing to try is just adding backoffs to those and seeing where that gets us.

That's probably just a band-aid though. The 2 underlying problems here are that delete-all shouldn't be such a common operation. Maybe `archive-all` would be a more s3 friendly alternative?

I think the `get-file` issue would perhaps be fixed by caching. But that one seems a little more suspect and maybe suggests a deeper problem in our system.

**Comment by jdoliner** [comment]
@sjezewski let's chat about this one on monday

**Comment by sjezewski** [comment]
Makes sense.

I'll run w the backoff code for deletion and adding support for `archive-all`

And I agree about caching ... that's a fine solution for the right problem, but I'd want to make sure we've isolated it to that type of problem

**Comment by sjezewski** [comment]
Some updates! We now have:

- the delete backoff code (in place on master)
- the new getfile cache (solves a similar but different issue) ... about to land
- and the logging update (in place on master) to help us identify the bug next time this happens

**Comment by sjezewski** [comment]
More updates:

We also added some profiling tools, and between profiling and looking at logs we identified a cascading bug.

This bug would be triggered on restart. 

In particular, we were indexing a blank commit, which caused an endless loop of `CreateJob()`s to be called that wouldn't finish (because the commit ID was blank), resulting in a new commit that would be started and never finished on each output repo. The jobs would try and restart using backoff, but the pipeline's input repo watcher would also create new jobs, making the effect compound. This resulted in about 1 new `CreateJob` / output commit per second. After a day of this, we see about 100k commits (that were never finished).

We believe this is the source of the 'hanging' symptom. This cascade could happen anytime there was a restart ... from either normal k8s rescheduling or if we truly were doing too many `GetFile()`s making a pod unresponsive, docker would kill the process/container and k8s would reschedule.

We have a fix we're testing now.

Between this fix and the caching for `GetFile()`s, we think the issue should go away. But I'll leave this open so that we can validate that once its deployed.
